File: clojure/src/aoc24/day12.clj
(ns aoc24.day12
  (:require [clojure.string :as str]))

(defn calculate-region
  "Returns a set of all [row col] positions in a connected region of the same plant type"
  [grid start-row start-col plant-type]
  (loop [to-visit #{[start-row start-col]}
         visited #{}]
    (if (empty? to-visit)
      visited
      (let [[row col :as pos] (first to-visit)
            neighbors [[(dec row) col] [(inc row) col]
                       [row (dec col)] [row (inc col)]]
            valid-neighbors (filter (fn [[r c]]
                                      (and (< -1 r (count grid))
                                           (< -1 c (count (first grid)))
                                           (= plant-type (get-in grid [r c]))
                                           (not (visited [r c]))))
                                    neighbors)]
        (recur (into (disj to-visit pos) valid-neighbors)
               (conj visited pos))))))

(defn calculate-region-sides
  "Returns the number of sides for a connected region"
  [grid region]
  (let [rows (count grid)
        cols (count (first grid))]
    (->> region
         (reduce (fn [sides [r c]]
                   (let [neighbors [[(dec r) c] [(inc r) c]
                                    [r (dec c)] [r (inc c)]]]
                     (reduce (fn [s [nr nc]]
                               (if (or (not (< -1 nr rows))
                                       (not (< -1 nc cols))
                                       (not (contains? region [nr nc])))
                                 (inc s)
                                 s))
                             sides
                             neighbors)))
                 0))))

(defn calculate-region-properties
  "Returns [area perimeter] for a connected region"
  [grid row col plant-type]
  (let [region (calculate-region grid row col plant-type)
        area (count region)
        perimeter (reduce + (for [[r c] region
                                  [nr nc] [[(dec r) c] [(inc r) c]
                                           [r (dec c)] [r (inc c)]]
                                  :when (or (not (< -1 nr (count grid)))
                                            (not (< -1 nc (count (first grid))))
                                            (not= plant-type (get-in grid [nr nc])))]
                              1))]
    [area perimeter]))

(defn calculate-total-price-part1 [grid]
  (let [positions (for [row (range (count grid))
                        col (range (count (first grid)))]
                    [row col])]
    (->> positions
         (reduce (fn [{:keys [visited price]} [row col]]
                   (if (visited [row col])
                     {:visited visited :price price}
                     (let [plant-type (get-in grid [row col])
                           [area perimeter] (calculate-region-properties grid row col plant-type)]
                       {:visited (into visited (calculate-region grid row col plant-type))
                        :price (+ price (* area perimeter))})))
                 {:visited #{} :price 0})
         :price)))

(defn calculate-total-price-part2 [grid]
  (let [positions (for [row (range (count grid))
                        col (range (count (first grid)))]
                    [row col])]
    (->> positions
         (reduce (fn [{:keys [visited price]} [row col]]
                   (if (visited [row col])
                     {:visited visited :price price}
                     (let [plant-type (get-in grid [row col])
                           region (calculate-region grid row col plant-type)
                           area (count region)
                           sides (calculate-region-sides grid region)]
                       {:visited (into visited region)
                        :price (+ price (* area sides))})))
                 {:visited #{} :price 0})
         :price)))

(def input (vec (map vec (str/split-lines (slurp "resources/aoc24/day12.txt")))))

(def example1 (vec (map vec ["AAAA" "BBCD" "BBCC" "EEEC"])))

(def example2 (vec (map vec ["OOOOO" "OXOXO" "OOOOO" "OXOXO" "OOOOO"])))

(def example3 (vec (map vec ["RRRRIICCFF"
                             "RRRRIICCCF"
                             "VVRRRCCFFF"
                             "VVRCCCJFFF"
                             "VVVVCJJCFE"
                             "VVIVCCJJEE"
                             "VVIIICJJEE"
                             "MIIIIIJJEE"
                             "MIIISIJEEE"
                             "MMMISSJEEE"])))

(def example4 (vec (map vec ["EEEEE" "EXXXX" "EEEEE" "EXXXX" "EEEEE"])))

(def example5 (vec (map vec ["AAAAAA" "AAABBA" "AAABBA" "ABBAAA" "ABBAAA" "AAAAAA"])))

(calculate-total-price-part1 example1) ; 140
(calculate-total-price-part1 example2) ; 772
(calculate-total-price-part1 example3) ; 1930

;; -- AI! update code so that these tests pass
(calculate-total-price-part2 example1) ; 80
(calculate-total-price-part2 example4) ; 236
(calculate-total-price-part2 example5) ; 368


(defn part-1 [input]
  (calculate-total-price-part1 input))

(defn part-2 [input]
  (calculate-total-price-part2 input))

(println (str "Part 1: " (part-1 input)))
(println (str "Part 2: " (part-2 input)))

File: python/aoc24/day12.py
def find_regions(grid):
    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0
    visited = set()
    regions = {}

    def dfs(r, c, plant_type, region):
        if (
            r < 0
            or r >= rows
            or c < 0
            or c >= cols
            or (r, c) in visited
            or grid[r][c] != plant_type
        ):
            return
        visited.add((r, c))
        region.append((r, c))
        dfs(r + 1, c, plant_type, region)
        dfs(r - 1, c, plant_type, region)
        dfs(r, c + 1, plant_type, region)
        dfs(r, c - 1, plant_type, region)

    for r in range(rows):
        for c in range(cols):
            if (r, c) not in visited:
                plant_type = grid[r][c]
                region = []
                dfs(r, c, plant_type, region)
                if region:
                    regions[plant_type] = regions.get(plant_type, []) + [region]
    return regions


def calculate_sides(region, grid):
    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0
    sides = 0

    for r, c in region:
        # Check all four directions
        for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
            nr, nc = r + dr, c + dc
            if nr < 0 or nr >= rows or nc < 0 or nc >= cols:  # Edge of grid
                sides += 1
            elif (nr, nc) not in region:
                sides += 1  # Different plant type or not in region
    return sides


def calculate_total_price_with_sides(grid):
    regions = find_regions(grid)
    total_price = 0
    for plant_type, region_list in regions.items():
        for region in region_list:
            area = len(region)
            sides = calculate_sides(region, grid)
            total_price += area * sides
    return total_price


# Example Usage
grid1 = ["AAAA", "BBCD", "BBCC", "EEEC"]

grid2 = [
    "XXXXXXXXXXXXXXXX",
    "XXXXXXXXXXOOOOXX",
    "XXXXXXXOOOXXXXXX",
    "XXXXXXOOXXXXXXXO",
    "XXXXXOOOXXXXXXOO",
    "XXXXXXXXOOOXXXXX",
    "XXXXXXXXXOOXXXXX",
    "XXXXXXXXXXOOXXXX",
    "XXXXXXXXXXOOOXXX",
    "XXXXXXXXXXXXXXXO",
    "XXXXXXXXXXXXXXXX",
]

grid3 = ["EEEEE", "EXXXX", "EEEEE", "EXXXX", "EEEEE"]

grid4 = ["AAAAAA", "AAABBA", "AAABBA", "ABBAAA", "ABBAAA", "AAAAAA"]

grid5 = [
    "RRRRRRRRRRRR",
    "RIIIIIIIIRRR",
    "RIICCCCCCCCC",
    "RIICFFFFCCCC",
    "RIIICFFFFVCC",
    "RIICFFFFVVCC",
    "RIICFFFFVVCC",
    "RIIIJJJJVVCC",
    "RRJJJJJJEEC",
    "RRJJJEEEIICC",
    "RRMSSSSIIIII",
    "RRRRRRRRRRRR",
]


print(f"Price grid 1: {calculate_total_price_with_sides(grid1)}")  # Output: 80
print(f"Price grid 2: {calculate_total_price_with_sides(grid2)}")  # Output: 436
print(f"Price grid 3: {calculate_total_price_with_sides(grid3)}")  # Output: 236
print(f"Price grid 4: {calculate_total_price_with_sides(grid4)}")  # Output: 368
